File: Graph/detect_cycle_dir_grph_dfs.java
import java.util.*;
class Solution {
    public boolean isCycleDfs(List<List<Integer>> adj, boolean[] vis, boolean[] inRecur, int node){
        vis[node]=true;
        inRecur[node]=true;
        for(int i: adj.get(node)){
            if(!vis[i] && isCycleDfs(adj,vis,inRecur,i))
            return true;
            else if(inRecur[i])
            return true;
        }
        inRecur[node]=false;
        return false;
    }
    public boolean isCyclic(int V, int[][] edges) {
        // code here
        boolean[] vis = new boolean[V];
        boolean[] inRecur = new boolean[V];
        List<List<Integer>> adj = new ArrayList<>();
        for(int i=0; i<V; i++){
            adj.add(new ArrayList<>());
        }
        for(int i=0; i<edges.length; i++){
            int u = edges[i][0];
            int v = edges[i][1];
            adj.get(u).add(v);
        }
        for(int i=0; i<V; i++){
            if(!vis[i] && isCycleDfs(adj,vis,inRecur,i))
            return true;
        }
        return false;
    }
}

File: Graph/detect_cycle_dir_grph_dfs.cpp
class Solution {
  private:
  bool isCycleDFS(vector<vector<int>> &adj, vector<bool> &vis, vector<bool> &inRecursion, int node){
      vis[node]=true;
      inRecursion[node]=true;
      for(auto &it: adj[node]){
          if(!vis[it] && isCycleDFS(adj,vis,inRecursion,it))
          return true;
          else if(inRecursion[it]) //matlab vis[it] aur inRecursion[it] dono true h
          return true;
      }
      inRecursion[node]=false;
      return false;
  }
  public:
    bool isCyclic(int V, vector<vector<int>> &edges) {
      vector<vector<int>> adj(V);
      for(int i=0; i<edges.size(); i++){
          int u=edges[i][0];
          int v= edges[i][1];
          adj[u].push_back(v);
      }
      vector<bool> vis(V,false);
      vector<bool> inRecursion(V,false);
      for(int i=0; i<V; i++){
          if(!vis[i] && isCycleDFS(adj,vis,inRecursion,i))
          return true;
      }
      return false;
    }
};